File: cobots_core/src/cobots_model/data/waypoint.py
from geometry import Pose, Position, Orientation


class Waypoint(Pose):

    '''
    Data structure methods
    '''

    def __init__(self, position=None, orientation=None, joints=None, type='',
                 name='', uuid=None, parent=None, append_type=True):
        super(Waypoint,self).__init__(
            position=position,
            orientation=orientation,
            type='waypoint.'+type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self._joints = None
        self.joints = joints

    def to_dct(self):
        msg = super(Waypoint,self).to_dct()
        msg.update({
            'joints': self.joints
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            position=Position.from_dct(dct['position']),
            orientation=Orientation.from_dct(dct['orientation']),
            type=dct['type'],
            append_type=False,
            name=dct['name'],
            uuid=dct['uuid'],
            joints=dct['joints'])

    '''
    Data accessor/modifier methods
    '''

    @property
    def joints(self):
        return self._joints

    @joints.setter
    def joints(self, value):
        if self._joints != value:
            self._joints = value
            if self._parent != None:
                self._parent.child_changed_event(
                    [self._child_changed_event_msg('joints','set')])

    def set(self, dct):
        joints = dct.get('joints',None)
        if joints != None:
            self.joints = joints

        super(Waypoint,self).set(dct)

File: evd_authoring_app/src/model/evdScript/data/waypoint.js
import { Pose, Position, Orientation } from './geometry';
import Blockly from 'blockly';

export class Waypoint extends Pose {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'waypoint.';
    }

    static fullTypeString() {
        return Pose.fullTypeString() + Waypoint.typeString();
    }

    constructor(position=null, orientation=null, joints=null, 
                type='', name='', uuid=null, parent=null, appendType=true) 
    {
        super(
            position,
            orientation,
            (appendType) ? 'waypoint.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );

        this._joints = null;

        this.joints = joints;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            joints: this.joints
        };
        return msg;
    }

    static fromDict(dct) {
        return new Waypoint(
            Position.fromDict(dct.position),
            Orientation.fromDict(dct.orientation),
            dct.joints,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    static BlocklyToolbox() {
        return { type: 'waypoint' };
    }

    static BlocklyBlock() {
        return { key: 'waypoint', data: {
            init: function() {
                this.appendDummyInput()
                    .appendField("Waypoint");
                this.appendDummyInput()
                    .appendField(new Blockly.FieldDropdown([["unnamed","default"]]), "name");
                this.setInputsInline(true);
                this.setPreviousStatement(true, ["trajectory", "waypoint"]);
                this.setNextStatement(true, ["trajectory", "waypoint"]);
                this.setColour(290);
                this.setTooltip("waypoint");
                this.setHelpUrl("waypoint");
            }
        }};
    }

    /*
    * Data accessor/modifier methods
    */

    get joints() {
        return this._joints;
    }

    set joints(value) {
        if (this._joints !== value) {
            this._joints = value;
            this.updatedAttribute('joints','set');
        }
    }

    set(dct) {

        if ('joints' in dct) {
            this.joints = dct.joints;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('joints','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('joints','update');
    }

}